lib/vast/Conversion/ABI/EmitABI.cpp

```diff
@@ -833,16 +826,20 @@ namespace vast
             mlir::ConversionTarget target(mctx);
             target.markUnknownOpDynamicallyLegal([](auto) { return true; });
 
-            auto should_transform = [&](hl::FuncOp op)
+            auto should_transform = [&](operation op)
             {
                 // TODO(conv:abi): We should always emit main with a fixed type.
-                return op.getName() == "main";
+                if (auto fn = mlir::dyn_cast< mlir::FunctionOpInterface >(op))
+                    return fn.getName() == "main";
+                return true;
             };
 
-            target.addDynamicallyLegalOp< hl::FuncOp >(should_transform);
+            target.markUnknownOpDynamicallyLegal(should_transform);
+            target.addLegalOp< abi::FuncOp >();
 
             mlir::RewritePatternSet patterns(&mctx);
             patterns.add< func_type< hl::FuncOp > >(tc, abi_info_map, mctx);
+            patterns.add< func_type< ll::FuncOp > >(tc, abi_info_map, mctx);
```

Comment:
Can not you generalize these patterns for FunctionTypeInterface or something like that?